File: java/client/src/org/openqa/selenium/json/Json.java
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.openqa.selenium.json;

import java.io.IOException;
import java.io.Reader;
import java.io.StringReader;
import java.io.StringWriter;
import java.io.UncheckedIOException;
import java.io.Writer;
import java.lang.reflect.Type;
import java.util.List;
import java.util.Map;

public class Json {
  public static final Type LIST_OF_MAPS_TYPE = new TypeToken<List<Map<String, Object>>>() {}.getType();
  public static final Type MAP_TYPE = new TypeToken<Map<String, Object>>() {}.getType();
  public static final Type OBJECT_TYPE = new TypeToken<Object>() {}.getType();

  private final JsonTypeCoercer fromJson = new JsonTypeCoercer();

  public String toJson(Object toConvert) {
    try (Writer writer = new StringWriter();
        JsonOutput jsonOutput = newOutput(writer)) {
      jsonOutput.write(toConvert);
      return writer.toString();
    } catch (IOException e) {
      throw new JsonException(e);
    }
  }

  public <T> T toType(String source, Type typeOfT) {
    return toType(source, typeOfT, PropertySetting.BY_NAME);
  }

  public <T> T toType(String source, Type typeOfT, PropertySetting setter) {
    try (StringReader reader = new StringReader(source)) {
      return toType(reader, typeOfT, setter);
    }
  }

  public <T> T toType(Reader source, Type typeOfT) {
    return toType(source, typeOfT, PropertySetting.BY_NAME);
  }

  public <T> T toType(Reader source, Type typeOfT, PropertySetting setter) {
    if (setter == null) {
      throw new JsonException("Mechanism for setting properties must be set");
    }

    try (JsonInput json = newInput(source)) {
      return fromJson.coerce(json, typeOfT, setter);
    }
  }

  public JsonInput newInput(Reader from) throws UncheckedIOException {
    return new JsonInput(from, fromJson, PropertySetting.BY_NAME);
  }

  public JsonOutput newOutput(Appendable to) throws UncheckedIOException {
    return new JsonOutput(to);
  }
}

File: javascript/atoms-ts/src/json.ts
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

/**
 * JSON utilities that leverage native JSON APIs available in all modern browsers.
 * All WebDriver-supported browsers have native JSON support, so we simply
 * use JSON.stringify and JSON.parse directly.
 *
 * @see https://caniuse.com/#search=JSON
 */

/**
 * Converts a JSON-serializable object to its string representation.
 *
 * @param jsonObj The input object to serialize
 * @param replacer Optional replacer function called for each (key, value) pair
 *        that determines how the value should be serialized
 * @returns A JSON string representation of the input object
 * @throws {TypeError} If the object contains non-serializable values
 *
 * @example
 * stringify({ name: 'Alice', age: 30 }) // '{"name":"Alice","age":30}'
 * stringify([1, 2, 3])                   // '[1,2,3]'
 */
export function stringify(
    jsonObj: unknown,
    replacer?: (key: string, value: unknown) => unknown
): string {
    return JSON.stringify(jsonObj, replacer as Parameters<typeof JSON.stringify>[1]);
}

/**
 * Parses a JSON string and returns the deserialized object.
 *
 * @param jsonStr The JSON string to parse
 * @param reviver Optional reviver function called for each (key, value) pair
 *        during deserialization
 * @returns The deserialized JSON object
 * @throws {SyntaxError} If the input string is not valid JSON
 *
 * @example
 * parse('{"name":"Alice","age":30}') // { name: 'Alice', age: 30 }
 * parse('[1,2,3]')                   // [1, 2, 3]
 */
export function parse(
    jsonStr: string,
    reviver?: (key: string, value: unknown) => unknown
): unknown {
    return JSON.parse(jsonStr, reviver as Parameters<typeof JSON.parse>[1]);
}
